### PR TITLE
Document HorizonClient and Models

### LINKED ISSUE #18: Document HorizonClient and Models
In order to break up the creation of documentation and avoid a ginormous pull request with documentation for the entire crate, break up documentation into clear bits:
* HorizonClient and models (this issue)
* Accounts and Assets (https://github.com/Baseflow/stellar-rust-sdk/issues/20)
* Claimable Balances and Ledgers

### PR DESCRIPTION
**Pull request by ejorritsma**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Introduces elaborate documentation for the HorizonClient and Models.

Furthermore:
* Moves HorizonClient into root, since it was the only struct in the module. This made for more natural documentation.
* Moves the XDR test into the Models testing suite.
* Removes the now empty horizon_client and xdr modules.

### :arrow_heading_down: What is the current behavior?
Currently the documentation is sparse and sometimes even inaccurate, making it hard for developers to understand how to use this SDK.

### :new: What is the new behavior (if this is a feature change)?
Elaborate and descriptive documentation.

### :boom: Does this PR introduce a breaking change?
no

### :bug: Recommendations for testing
`cargo doc --no-deps --open`

if you also want to see documentation for private items:
`cargo doc --no-deps --document-private-items --open`

### :memo: Links to relevant issues/docs
https://github.com/Baseflow/stellar-rust-sdk/issues/18

### :thinking: Checklist before submitting

- [x] All projects build
- [x] Follows style guide lines 
- [x] Relevant documentation was updated
- [x] Rebased onto current main

### CODE REVIEW

File: src/horizon_client.rs [hunk 1]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
+    /// # use stellar_rust_sdk::accounts::prelude::*;
+    /// # use stellar_rust_sdk::accounts::accounts_request::Signer;
+    /// # use stellar_rust_sdk::models::Request;
+    /// # use stellar_rust_sdk::horizon_client::HorizonClient;
+    /// # 
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let base_url = "https://horizon-testnet.stellar.org".to_string();
+    /// # let horizon_client = HorizonClient::new(base_url)
+    /// #    .expect("Failed to create Horizon Client");
+    /// let request: AccountsRequest<NoSponsor, Signer, NoAsset, NoLiquidityPool> = AccountsRequest::new()
```
Comment (tluijken) [inline src/horizon_client.rs]: I wonder if people would annotate their variable like this. Perhaps to not make it too complicated, just stick with 
```suggestion
    /// let request = AccountsRequest::new()
```

File: src/horizon_client.rs [hunk 2]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
+    /// which encapsulates the list of accounts retrieved from the server. 
+    /// In case of a failure in the request, it returns an error encapsulated within `Result`.
+    ///
+    /// # Example
+    /// To use this method, create an instance of [`AccountsRequest`] and set at least
+    /// one of the four filter options. For example, filtering by signer:
+    /// 
+    /// ```rust
+    /// # use stellar_rust_sdk::accounts::prelude::*;
+    /// # use stellar_rust_sdk::accounts::accounts_request::Signer;
```
Comment (tluijken) [inline src/horizon_client.rs]: If we can agree on the suggestion below, we might not have to add the `Signer` using, and just stick with the `prelude`.

File: src/horizon_client.rs [hunk 3]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
     /// # Returns
-    /// The all assets response
-    /// # Errors
-    /// Returns an error if the request fails
-    /// [GET /assets](https://www.stellar.org/developers/horizon/reference/endpoints/assets-all.html)
+    ///
+    /// On success, this method returns a `Result` wrapping an [`AllAssetsResponse`], which includes 
+    /// the comprehensive list of assets retrieved from the Horizon server. If the request 
+    /// encounters an issue, an error is returned within `Result`.
+    ///
```
Comment (tluijken) [inline src/horizon_client.rs]: We should add en example here too, just like the `get_single_account` request.

File: src/horizon_client.rs [hunk 4]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
     /// # Returns
-    /// The all claimable balances response
-    /// # Errors
-    /// Returns an error if the request fails
-    /// [GET /claimable_balances/all](https://www.stellar.org/developers/horizon/reference/endpoints/claimable_balances-all.html)
+    ///
+    /// Returns a `Result` containing an [`AllClaimableBalancesResponse`] with the list of all 
+    /// claimable balances if successful. In case of a failure, it returns an error within `Result`.
+    ///
     pub async fn get_all_claimable_balances(
```
Comment (tluijken) [inline src/horizon_client.rs:226]: We should add en example here too, just like the `get_single_account` request.

File: src/horizon_client.rs [hunk 5]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
     /// # Returns
-    /// The ledgers response
-    /// # Errors
-    /// Returns an error if the request fails
-    /// [GET /ledgers](https://www.stellar.org/developers/horizon/reference/endpoints/ledgers-all.html)
+    ///
+    /// On successful execution, returns a `Result` containing a [`LedgersResponse`], 
+    /// which includes the list of all ledgers obtained from the Horizon server. If the request 
+    /// fails, it returns an error within `Result`.
+    ///
```
Comment (tluijken) [inline src/horizon_client.rs:320]: We should add en example here too, just like the `get_single_account` request.

File: src/horizon_client.rs [hunk 6]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
+    /// ```
+    /// # use stellar_rust_sdk::claimable_balances::prelude::*;
+    /// # use stellar_rust_sdk::models::Request;
+    /// # use stellar_rust_sdk::horizon_client::HorizonClient;
+    /// # 
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let base_url = "https://horizon-testnet.stellar.org".to_string();
+    /// # let horizon_client = HorizonClient::new(base_url)
+    /// #    .expect("Failed to create Horizon Client");
+    ///  let request: SingleClaimableBalanceRequest<ClaimableBalanceId> = SingleClaimableBalanceRequest::new()
```
Comment (tluijken) [inline src/horizon_client.rs]: We should probably omit the annotation, as we want our states to eventually become internal

File: src/horizon_client.rs [hunk 7]
```diff
@@ -17,183 +14,434 @@ use reqwest;
...
+    /// # use stellar_rust_sdk::accounts::prelude::*;
+    /// # use stellar_rust_sdk::accounts::accounts_request::Signer;
+    /// # use stellar_rust_sdk::models::Request;
+    /// # use stellar_rust_sdk::horizon_client::HorizonClient;
+    /// # 
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let base_url = "https://horizon-testnet.stellar.org".to_string();
+    /// # let horizon_client = HorizonClient::new(base_url)
+    /// #    .expect("Failed to create Horizon Client");
+    /// let request: SingleAccountRequest<AccountId> = SingleAccountRequest::new()
```
Comment (tluijken) [inline src/horizon_client.rs]: We should probably omit the annotation, as we want our states to eventually become internal

File: src/lib.rs [hunk 8]
```diff
@@ -113,21 +207,53 @@ impl std::fmt::Display for AssetType {
...
 trait BuildQueryParametersExt<T> {
+    /// Constructs a query string for an HTTP request from the object's properties.
+    ///
+    /// This method transforms the properties of the implementing object into a URL-encoded query 
+    /// string. Each property is converted to a key-value pair, and pairs are concatenated with '&'.
+    /// Properties that are `None` are omitted from the string.
+    ///
+    /// # Returns
+    /// A `String` representing the query parameters of the HTTP request. If there are no 
+    /// parameters, or all properties are `None`, an empty string is returned.
```
Comment (tluijken) [inline src/lib.rs]: Dit laatste is alleen waar voor de implementation van `Vec<Option<T>>`. Voor de algemene generic is dit niet persee waar.

**Review by tluijken — CHANGES_REQUESTED** [review]

File: src/horizon_client.rs (hunk 2 above)
Comment (ejorritsma) [reply src/horizon_client.rs]: Removed the Signer import :+1:

File: src/horizon_client.rs (hunk 3 above)
Comment (ejorritsma) [reply src/horizon_client.rs]: Added an example!

File: src/horizon_client.rs (hunk 4 above)
Comment (ejorritsma) [reply src/horizon_client.rs:226]: Added an example!

File: src/horizon_client.rs (hunk 5 above)
Comment (ejorritsma) [reply src/horizon_client.rs:320]: Added an example!

File: src/horizon_client.rs (hunk 6 above)
Comment (ejorritsma) [reply src/horizon_client.rs]: Removed the annotation :+1:

File: src/horizon_client.rs (hunk 7 above)
Comment (ejorritsma) [reply src/horizon_client.rs]: Removed the annotation :+1:

File: src/lib.rs (hunk 8 above)
Comment (ejorritsma) [reply src/lib.rs]: Ik heb nu deze subtlety (die wel belangrijk is) verwerkt in de documentatie zodat duidelijk is dat deze implementatiedetails/return contents alleen gelden bij de specifieke implementatie voor `Vec<Option<T>>`. Ik denk dat het zo beter is, misschien kun je er even naar kijken? :+1:

File: src/lib.rs [hunk 9]
```diff
@@ -1,103 +1,197 @@
...
+/// let request = AccountsRequest::new()
+///     .set_signer("GDQJUTQYK2MQX2VGDR2FYWLIYAQIEGXTQVTFEMGH2BEWFG4BRUY4CKI7").unwrap()
+///     .set_limit(10).unwrap();
+///
+/// let response: Result<AccountsResponse, String> = horizon_client
+///     .get_account_list(&request)
+///     .await;
+/// # Ok({})
+/// # }
+///
```
Comment (tluijken) [inline src/lib.rs:118]: Does this documentation belong to the accounts module? I believe this exact same documentation is already provided at https://github.com/Baseflow/stellar-rust-sdk/blob/4cf167e4dee31fdf8b9f31895c0b6a12dd9a2235/src/horizon_client/horizon_client.rs#L44. The module itself contains of course more than the provided documentation here.

File: src/lib.rs [hunk 10]
```diff
@@ -1,103 +1,197 @@
...
+/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+/// # let base_url = "https://horizon-testnet.stellar.org".to_string();
+/// # let horizon_client = HorizonClient::new(base_url)
+/// #    .expect("Failed to create Horizon Client");;
+/// let all_assets_request = AllAssetsRequest::new();
+/// let accounts_response: Result<AllAssetsResponse, String> = horizon_client
+///     .get_all_assets(&all_assets_request)
+///     .await;
+/// # Ok({})
+/// # }
```
Comment (tluijken) [inline src/lib.rs:156]: This also has been documentated at HorizonClient::getAllAssets. The module itself provides more functionality, so it should not be documented here for a specific endpoint I believe.

File: src/lib.rs [hunk 11]
```diff
@@ -1,103 +1,197 @@
...
+/// use stellar_rust_sdk::horizon_client::HorizonClient;
+///
+/// let base_url = "https://horizon-testnet.stellar.org".to_string();
+/// let horizon_client = HorizonClient::new(base_url)
+///     .expect("Failed to create Horizon Client");;
+/// ```
+///
+/// # Using the `HorizonClient`
+/// The HorizonClient has a function that can be called for each endpoind provided
+/// by the Horizon API. For example, it has a [`HorizonClient::get_account_list`](crate::horizon_client::HorizonClient::get_account_list)
```
Comment (tluijken) [inline src/lib.rs:140]: This refers to the get_account_list fn, while the example shows the AllAssetsRequest example.

**Review by tluijken — CHANGES_REQUESTED** [review]
I think we can omit the code examples on the mod exports. For specific calls, the horizonClient already provides the same examples per function.

File: src/lib.rs (hunk 8 above)
Comment (tluijken) [reply src/lib.rs]: Ik zou m verplaatsen naar de documentatie van de implementatie. De filtering is specifiek voor die implementatie, maar de trait dwingt het niet af.

**Review by tluijken — CHANGES_REQUESTED** [review]
Ik neem die documentatie nog even over

**Comment by tluijken** [comment]
Already processed with PR #21

**Pull request closed without merging**